src/main/java/minijava/ir/optimize/Optimizer.java

```diff
@@ -52,23 +52,26 @@ static void optimize() {
             .dependsOn(controlFlowOptimizer, floatInTransformation)
             .build();
 
+    ProgramMetrics metrics = ProgramMetrics.analyse(firm.Program.getGraphs());
+    Inliner inliner = new Inliner(metrics, true);
     while (true) {
       for (Graph graph : firm.Program.getGraphs()) {
         perGraphFramework.optimizeUntilFixedpoint(graph);
       }
 
       // Here comes the interprocedural stuff... This is method is really turning into a mess
       boolean hasChanged = false;
-      ProgramMetrics metrics = ProgramMetrics.analyse(firm.Program.getGraphs());
-      Optimizer inliner = new Inliner(metrics);
       for (Graph graph : firm.Program.getGraphs()) {
         hasChanged |= inliner.optimize(graph);
         unreachableCodeRemover.optimize(graph);
-        metrics.updateGraphInfo(graph);
         Cli.dumpGraphIfNeeded(graph, "after-inlining");
       }
       if (!hasChanged) {
-        break;
+        if (inliner.onlyLeafs) {
+          inliner = new Inliner(metrics, false);
```

Comment:
I didn't say I didn't want object-orientation, I said that I really didn't want to have abstract classes and that instead we should extract an interface.

In this case the interface we'd extract would have exactly one method: `boolean inlineOnlyLeafs();`, where it's two implementations would return `false` and `true` respectively.

You are right in that this is *exactly* like an `enum`, so that's what I suggested.